### PR TITLE
pml_ucx: implement cancel, and add small optimizations.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by yosefe** [comment]
@shamisp @miked-mellanox @bosilca

File: ompi/mca/pml/ucx/pml_ucx_request.c [hunk 1]
```diff
@@ -55,12 +61,19 @@ void mca_pml_ucx_recv_completion(void *request, ucs_status_t status,
     OPAL_THREAD_UNLOCK(&ompi_request_lock);
 }
 
+static void mca_pml_ucx_persistent_requset_detach(mca_pml_ucx_persistent_request_t *preq,
```
Comment (bosilca) [inline ompi/mca/pml/ucx/pml_ucx_request.c]: Isn't that a typo (requset)?

File: ompi/mca/pml/ucx/pml_ucx_request.c [hunk 2]
```diff
@@ -55,12 +61,19 @@ void mca_pml_ucx_recv_completion(void *request, ucs_status_t status,
...
+    preq->tmp_req                 = NULL;
+}
+
 void mca_pml_ucx_persistent_requset_complete(mca_pml_ucx_persistent_request_t *preq,
                                              ompi_request_t *tmp_req)
 {
     preq->ompi.req_status = tmp_req->req_status;
     ompi_request_complete(&preq->ompi, true);
-    tmp_req->req_complete_cb_data = NULL;
+    mca_pml_ucx_persistent_requset_detach(preq, tmp_req);
```
Comment (bosilca) [inline ompi/mca/pml/ucx/pml_ucx_request.c]: propagated here.

File: ompi/mca/pml/ucx/pml_ucx_request.c [hunk 3]
```diff
@@ -133,18 +148,35 @@ void mca_pml_ucx_request_cleanup(void *request)
 
 static int mca_pml_ucx_persistent_request_free(ompi_request_t **rptr)
 {
-    mca_pml_ucx_persistent_request_t* req = (mca_pml_ucx_persistent_request_t*)*rptr;
+    mca_pml_ucx_persistent_request_t* preq = (mca_pml_ucx_persistent_request_t*)*rptr;
+    ompi_request_t *tmp_req = preq->tmp_req;
 
+    preq->ompi.req_state = OMPI_REQUEST_INVALID;
+    if (tmp_req != NULL) {
+        mca_pml_ucx_persistent_requset_detach(preq, tmp_req);
```
Comment (bosilca) [inline ompi/mca/pml/ucx/pml_ucx_request.c]: and here.

**Comment by bosilca** [comment]
Other than the typo the patch looks good. :+1:

**Comment by shamisp** [comment]
Agree with George's comments. +1

**Comment by yosefe** [comment]
fixed the typo.

**Pull request merged**